spec/forms/reactivate_profile_form_spec.rb

```diff
@@ -0,0 +1,93 @@
+require 'rails_helper'
+
+describe ReactivateProfileForm do
+  subject(:form) do
+    ReactivateProfileForm.new(user,
+                              recovery_code: recovery_code,
+                              password: password)
+  end
+
+  let(:user) { build_stubbed(:user) }
+  let(:recovery_code) { nil }
+  let(:password) { nil }
+
+  describe '#valid?' do
+    let(:password) { 'asd' }
+    let(:recovery_code) { '123' }
+    let(:valid_recovery_code?) { true }
+    let(:valid_password?) { true }
+
+    before do
+      allow(form).to receive(:valid_password?).and_return(valid_password?)
+      allow(form).to receive(:valid_recovery_code?).and_return(valid_recovery_code?)
+    end
+
+    context 'when required attributes are not present' do
+      let(:password) { nil }
+      let(:recovery_code) { nil }
+
+      it 'is invalid' do
+        expect(subject).to_not be_valid
+        expect(subject.errors[:recovery_code]).to eq [t('errors.messages.blank')]
+        expect(subject.errors[:password]).to eq [t('errors.messages.blank')]
+      end
+    end
+
+    context 'when there is no profile that has had its password reset' do
+      let(:password_reset_profile) { nil }
+
+      it 'is invalid' do
+        expect(subject).to_not be_valid
+        expect(subject.errors[:base]).to eq [t('errors.messages.no_password_reset_profile')]
+      end
+    end
+
+    context 'when recovery code does not match' do
+      let(:valid_recovery_code?) { false }
+
+      it 'is invalid' do
+        expect(subject).to_not be_valid
+        expect(subject.errors[:recovery_code]).to eq [t('errors.messages.recovery_code_incorrect')]
+      end
+    end
+
+    context 'when password does not match' do
+      let(:valid_password?) { false }
+
+      it 'is invalid' do
+        expect(subject).to_not be_valid
+        expect(subject.errors[:password]).to eq [t('errors.messages.password_incorrect')]
+      end
+    end
+  end
+
+  describe '#submit' do
+    before { expect(form).to receive(:valid?).and_return(valid?) }
+
+    let(:flash) { {} }
+    subject(:submit) { form.submit(flash) }
```

Comment:
can we use vars instead of `subject` ? I think the usage here points to why it is confusing -- we are nesting `subject` so `it` is something different depending on the spec you are reading